### PR TITLE
8266461: tools/jmod/hashes/HashesTest.java fails: static @Test methods

### PR DESCRIPTION
**Pull request by LanceAndersen**
HI all,

Please review the fix to HashesTest.java to support running on TestNG 7.4.  

The fix adds a no-arg constructor which TestNG requires.

The change allows the test to run with the current jtreg as well as the upcoming jtreg

Best
Lance

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266461](https://bugs.openjdk.java.net/browse/JDK-8266461): tools/jmod/hashes/HashesTest.java fails: static @Test methods

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 2e8f6c97bd642e707bcab2e76302f645217f6cad
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 2e8f6c97bd642e707bcab2e76302f645217f6cad

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4009/head:pull/4009` \
`$ git checkout pull/4009`

Update a local copy of the PR: \
`$ git checkout pull/4009` \
`$ git pull https://git.openjdk.java.net/jdk pull/4009/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4009`

View PR using the GUI difftool: \
`$ git pr show -t 4009`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4009.diff">https://git.openjdk.java.net/jdk/pull/4009.diff</a>

</details>

### CODE REVIEW

File: test/jdk/tools/jmod/hashes/HashesTest.java [hunk 1]
```diff
@@ -86,6 +86,12 @@
     private final Path lib;
     private final ModuleInfoMaker builder;
 
+    public HashesTest() {
+        mods= null;
+        srcDir=null;
+        lib= null;
+        builder=null;
+    }
```
Comment (AlanBateman) [inline test/jdk/tools/jmod/hashes/HashesTest.java:94]: This looks like a workaround. Can you instead see if the fields can be changed to non-final and place the constructor with a method that has the `@BeforeClass` annotation?

**Review by AlanBateman — CHANGES_REQUESTED** [review]

**Comment by ChrisHegarty** [comment]
The non-static state in this test class is initialized for each of the static testXXX scenarios. An alternative could be to move said state (four fields) into a static inner class, and have each of the testXXX scenarios create an instance of that class with the test-specific path. That would also allow the addition of the no-args public constructor to HashesTest, and the testXXX methods to be made non-static.

**Comment by LanceAndersen** [comment]
> The non-static state in this test class is initialized for each of the static testXXX scenarios. An alternative could be to move said state (four fields) into a static inner class, and have each of the testXXX scenarios create an instance of that class with the test-specific path. That would also allow the addition of the no-args public constructor to HashesTest, and the testXXX methods to be made non-static.

I had originally thought about introducing an inner class but decided that given TestNG needs access to  the default constructor, I chose that route vs. doing more of an update.  I can revisit this though

File: test/jdk/tools/jmod/hashes/HashesTest.java (hunk 1 above)
Comment (LanceAndersen) [reply test/jdk/tools/jmod/hashes/HashesTest.java:94]: I can look to update the test.  As I mentioned in my reply to Chris, I had thought about introducing an inner class bug decided to go the route of the default/no-arg constructor as it required fewer changes.

**Comment by mlchung** [comment]
I also think it's good to fix this properly.    Each test wants to run in an unique directory.    Another solution is to make the fields non-final and add a new `@BeforeMethod` method to generate the unique pathname for each test.   I think this seems a clean way (I sent you a patch to checkout).

**Comment by LanceAndersen** [comment]
The latest commit takes into account the feedback received

> I also think it's good to fix this properly. Each test wants to run in an unique directory. Another solution is to make the fields non-final and add a new `@BeforeMethod` method to generate the unique pathname for each test. I think this seems a clean way (I sent you a patch to checkout).

The latest commit leverages @BeforeMethod vs an inner class and also includes some additional minor cleanup

File: test/jdk/tools/jmod/hashes/HashesTest.java [hunk 2]
```diff
@@ -81,17 +74,20 @@
...
-    HashesTest(Path dest) throws IOException {
+    @BeforeMethod
+    public void setTestPath() throws IOException {
+        Path dest = Path.of("test" + counter.addAndGet(1));
         if (Files.exists(dest)) {
             deleteDirectory(dest);
         }
         this.mods = dest.resolve("mods");
-        this.srcDir = dest.resolve("src");
+        Path srcDir = dest.resolve("src");
```
Comment (mlchung) [inline test/jdk/tools/jmod/hashes/HashesTest.java:90]: you can just get rid of this local variable and do this in line 92:
```
     this.builder = new ModuleInfoMaker(dest.resolve("src"));
```

File: test/jdk/tools/jmod/hashes/HashesTest.java [hunk 3]
```diff
@@ -401,19 +378,16 @@ private ModuleHashes hashes(String name) {
...
-            try (InputStream in = reader.open("module-info.class").get()) {
+            try (ModuleReader reader = mref.open(); InputStream in = reader.open("module-info.class").get()) {
                 ModuleHashes hashes = ModuleInfo.read(in, null).recordedHashes();
                 System.out.format("hashes in module %s %s%n", name,
-                    (hashes != null) ? "present" : "absent");
+                        (hashes != null) ? "present" : "absent");
                 if (hashes != null) {
                     hashes.names().stream().sorted().forEach(n ->
-                        System.out.format("  %s %s%n", n, toHex(hashes.hashFor(n)))
+                            System.out.format("  %s %s%n", n, toHex(hashes.hashFor(n)))
```
Comment (mlchung) [inline test/jdk/tools/jmod/hashes/HashesTest.java:387]: Nit: the extra whitespaces in line 384 and 387 may be added by IDE.  Can you revert it.

**Review by mlchung — COMMENTED** [review]
Looks like the imports and whitespace are changed by IDE??   can you please revert these unintentional change.

Otherwise, the change looks good.

File: test/jdk/tools/jmod/hashes/HashesTest.java (hunk 2 above)
Comment (LanceAndersen) [reply test/jdk/tools/jmod/hashes/HashesTest.java:90]: Yes I can do that and realized that after I pushed the update

File: test/jdk/tools/jmod/hashes/HashesTest.java (hunk 3 above)
Comment (LanceAndersen) [reply test/jdk/tools/jmod/hashes/HashesTest.java:387]: Intellij did that.  I can tweak

**Review by LanceAndersen — COMMENTED** [review]
I can look to revert the imports, that was an optimization via Intellij

**Review by mlchung — APPROVED** [review]

File: test/jdk/tools/jmod/hashes/HashesTest.java [hunk 4]
```diff
@@ -401,8 +386,7 @@ private ModuleHashes hashes(String name) {
     private ModuleHashes hashes(ModuleFinder finder, String name) {
         ModuleReference mref = finder.find(name).orElseThrow(RuntimeException::new);
         try {
-            ModuleReader reader = mref.open();
-            try (InputStream in = reader.open("module-info.class").get()) {
+            try (ModuleReader reader = mref.open(); InputStream in = reader.open("module-info.class").get()) {
```
Comment (AlanBateman) [inline test/jdk/tools/jmod/hashes/HashesTest.java:389]: Thats for doing the right thing and using BeforeMethod. The changes looks good.
There is some random re-formatting in hashes and deleteDirectory, I don't know if this was intended or not. I think we should at least resolve L389 before integrating.

**Review by AlanBateman — APPROVED** [review]

File: test/jdk/tools/jmod/hashes/HashesTest.java (hunk 4 above)
Comment (LanceAndersen) [reply test/jdk/tools/jmod/hashes/HashesTest.java]: Thank you Alan.  The re-formatting was done automagically by Intellij.  I just pushed an update to address your suggestions.

**Pull request closed without merging**